File: QViz.NET/Periscope.QViz/Clients/QVizClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using Microsoft.Azure.Storage;
using Microsoft.Azure.Storage.Blob;
using Newtonsoft.Json;
using Periscope.QViz.JSON;

namespace Periscope.QViz.Clients
{
	/// <summary>
	/// Client Library for interfacing with QViz APIs
	/// </summary>
	public class QVizClient
	{
		#region Private Properties

		private readonly Uri _baseURL;
		private AccessUser _user;
		private string _lastResponse;

		#endregion

		#region Constructors

		/// <summary>
		/// Initialize QViz Client with the default Base URL for the API Instance of QViz
		/// </summary>
		public QVizClient()
		{
			_baseURL = new Uri("https://api.qviz.io");
		}

		/// <summary>
		/// Initialize QViz Client with the provided Base URL for the API Instance of QViz
		/// </summary>
		/// <param name="baseURL">Base URL to the API Instance of QViz</param>
		public QVizClient(string baseURL)
		{
			_baseURL = new Uri(baseURL);
		}

		#endregion

		#region Public Properties

		/// <summary>
		/// Get the Response from the QViz API that was executed last
		/// </summary>
		/// <returns>Text content of the Response from the QViz API that was executed last</returns>
		public string LastResponse()
		{
			return _lastResponse;
		}

		/// <summary>
		/// Authenticate with QViz using the provided User credentials
		/// </summary>
		/// <param name="userName">Name of the User</param>
		/// <param name="password">Password of the User</param>
		/// <exception cref="Exception">In case of failure</exception>
		public void Authenticate(string userName, string password)
		{
			var authentication = new Authentication(userName, password);
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Body = authentication;
			_user = apiClient.Post<AccessUser>("/api/authentication");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		/// <summary>
		/// Get the list of Projects associated with the User
		/// </summary>
		/// <returns>List of Projects as Array of Objects</returns>
		/// <exception cref="Exception">In case of failure</exception>
		public List<Project> GetProjects()
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("userId", _user.userId));
			var projects = apiClient.Get<List<Project>>("/api/projects");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
			return projects;
		}

		/// <summary>
		/// Get a specific Project associated with the User
		/// </summary>
		/// <param name="projectName">Name of the Project to be fetched</param>
		/// <returns>Project Object</returns>
		public Project GetProject(string projectName)
		{
			var projects = GetProjects();
			return projects.FirstOrDefault(
				project => project.projectName.Equals(projectName, StringComparison.CurrentCultureIgnoreCase));
		}

		/// <summary>
		/// Get the list of Configurations associated with the give Project
		/// </summary>
		/// <param name="projectId">Identifier of the Project to get Configurations from</param>
		/// <returns>List of Configurations for the given Project</returns>
		public List<ProjectConfiguration> GetProjectConfigurations(string projectId)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Headers.Add(new KeyValue("userId", _user.userId));
			apiClient.Queries.Add(new KeyValue("projectId", projectId));
			var configurations = apiClient.Get<List<ProjectConfiguration>>("/api/projectConfigurations");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
			return configurations;
		}

		/// <summary>
		/// Get a specific Tool Configuration from the Project Configurations
		/// </summary>
		/// <param name="projectId">Identifier of the Project</param>
		/// <param name="toolName">Name of the Tool (SLACK, JIRA, TESTRAIL, etc.)</param>
		/// <typeparam name="T">Type of the JSON Model for Tool Configuration</typeparam>
		/// <returns>Tool Configuration Details as given JSON Model</returns>
		public T GetToolConfiguration<T>(string projectId, string toolName) where T : new()
		{
			var projectConfigurations = GetProjectConfigurations(projectId);
			foreach (var configuration in projectConfigurations)
			{
				if (configuration.value.IndexOf(toolName, StringComparison.Ordinal) > 0)
				{
					return JsonConvert.DeserializeObject<T>(configuration.value);
				}
			}

			// Throw an Exception if the given Tool Name was not found
			throw new Exception(
				"Project (" + projectId + ") does not contain a Tool Configuration for (" + toolName + ") tool."
			);
		}

		/// <summary>
		/// Get the list of GUI Test Cases associated with the specified Project
		/// </summary>
		/// <param name="projectId">Identifier for the Project to fetch GUI Test Cases from</param>
		/// <param name="onlyAutomated">Get only the GUI Test Cases that are Automated?</param>
		/// <param name="onlyDescription">Get only the GUI Test Case(s) that matches the Description?</param>
		/// <returns>List of GUI Test Cases as Array of Objects</returns>
		/// <exception cref="Exception">In case of failure</exception>
		public List<TestCaseGUI> GetGUITestCases(
			string projectId,
			bool onlyAutomated = false,
			string onlyDescription = "")
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("projectId", projectId));
			var guiCases = apiClient.Get<List<TestCaseGUI>>("/api/uiTestcases");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
			// Filter only Automated Test Cases (if requested)
			if (onlyAutomated)
			{
				guiCases = guiCases.Where(tc => tc.isAutomated.Equals(true)).ToList();
			}

			// Filter only Test Cases that matches the Description (if requested)
			if (!string.IsNullOrEmpty(onlyDescription))
			{
				guiCases = guiCases.Where(
					tc => tc.description.Equals(onlyDescription, StringComparison.CurrentCultureIgnoreCase)).ToList();
			}

			return guiCases;
		}

		/// <summary>
		/// Get the list of API Test Cases associated with the specified Project
		/// </summary>
		/// <param name="projectId">Identifier for the Project to fetch API Test Cases from</param>
		/// <param name="onlyAutomated">Get only the API Test Cases that are Automated?</param>
		/// <param name="onlyDescription">Get only the API Test Case(s) that matches the Description?</param>
		/// <returns>List of API Test Cases as Array of Objects</returns>
		/// <exception cref="Exception">In case of failure</exception>
		public List<TestCaseAPI> GetAPITestCases(
			string projectId,
			bool onlyAutomated = false,
			string onlyDescription = "")
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("projectId", projectId));
			var apiCases = apiClient.Get<List<TestCaseAPI>>("/api/apiTestcases");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
			// Filter only Automated Test Cases (if requested)
			if (onlyAutomated)
			{
				apiCases = apiCases.Where(tc => tc.isAutomated.Equals(true)).ToList();
			}

			// Filter only Test Cases that matches the Description (if requested)
			if (!string.IsNullOrEmpty(onlyDescription))
			{
				apiCases = apiCases.Where(
					tc => tc.description.Equals(onlyDescription, StringComparison.CurrentCultureIgnoreCase)).ToList();
			}

			return apiCases;
		}

		/// <summary>
		/// Upload the given Screen Shot File into Azure BLOB Storage and return the URL
		/// </summary>
		/// <param name="connection">Azure Storage Account Connection String</param>
		/// <param name="container">Azure Storage Container Name</param>
		/// <param name="file">Screen Shot File</param>
		/// <returns>Full URL of the Screen Shot File from Azure BLOB Storage</returns>
		/// <exception cref="Exception">In case of failure</exception>
		public static string UploadScreenShot(string connection, string container, FileInfo file)
		{
			try
			{
				// Make sure the File exists
				if (file is null || !file.Exists)
				{
					throw new Exception("Provided Screen Shot File does not exist.");
				}

				// Upload the File to BLOB and return the URL
				var storageAccount = CloudStorageAccount.Parse(connection);
				var blobClient = storageAccount.CreateCloudBlobClient();
				var blobContainer = blobClient.GetContainerReference(container);
				var blob = blobContainer.GetBlockBlobReference(file.Name);
				blob.UploadFromFile(file.FullName);
				return blob.Uri.AbsoluteUri;
			}
			catch (Exception)
			{
				return "";
			}
		}

		/// <summary>
		/// Post the provided GUI Testing Results to the QViz Instance
		/// </summary>
		/// <param name="guiTestResult">GUI Test Results as JSON Object</param>
		public void PostGUITestResults(GUITestResult guiTestResult)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Headers.Add(new KeyValue("userId", _user.userId));
			apiClient.Body = guiTestResult;
			apiClient.Post<object>("/api/ui-results");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		/// <summary>
		/// Post the provided API Testing Results to the QViz Instance
		/// </summary>
		/// <param name="apiTestResult">API Test Results as JSON Object</param>
		public void PostAPITestResults(APITestResult apiTestResult)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Headers.Add(new KeyValue("userId", _user.userId));
			apiClient.Body = apiTestResult;
			apiClient.Post<object>("/api/api-results");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		/// <summary>
		/// Send the provided GUI/API Testing Results to the Test Rail Instance configured with QViz
		/// </summary>
		/// <param name="testResult">GUI or API Test Results as JSON Object</param>
		/// <param name="testRailURL">URL of the TestRail Instance</param>
		/// <param name="userName">Unique Name of the TestRail User</param>
		/// <param name="apiToken">API Access Key of the TestRail User</param>
		/// <param name="testRailRunID">Identifier for the Test Run from TestRail Instance</param>
		/// <param name="lastError">Last Error Message (if any)</param>
		public void SendTestResultsToTestRail(
			object testResult,
			string testRailURL,
			string userName,
			string apiToken,
			string testRailRunID,
			string lastError)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("TestRailURL", testRailURL));
			apiClient.Queries.Add(new KeyValue("username", userName));
			apiClient.Queries.Add(new KeyValue("passwordKey", apiToken));
			apiClient.Queries.Add(new KeyValue("postResult", "true"));
			apiClient.Queries.Add(new KeyValue("TestRailRunID", testRailRunID));
			if (lastError != null && lastError.StartsWith("Defect", StringComparison.CurrentCultureIgnoreCase))
			{
				apiClient.Queries.Add(
					new KeyValue(
						"Defect",
						lastError.Split(": ".ToCharArray())[1].Split("::".ToCharArray())[0]
							.Replace(" :", "")
					)
				);
			}

			apiClient.Body = testResult;
			apiClient.Post<object>("/api/integration/TestRail");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		/// <summary>
		/// Send the provided GUI/API Testing Results to the Slack Instance configured with QViz
		/// </summary>
		/// <param name="testResult">GUI or API Test Results as JSON Object</param>
		/// <param name="slackURL">URL of the Slack Hooks for sending Messages</param>
		/// <param name="projectID">Identifier for the Project</param>
		/// <param name="projectName">Name of the Project</param>
		/// <param name="projectType">Type of the Project</param>
		/// <param name="qvizURL">URL to the QViz App Instance</param>
		/// <param name="runID">Identifier for the Run</param>
		public void SendTestResultsToSlack(
			object testResult,
			string slackURL,
			string projectID,
			string projectName,
			string projectType,
			string qvizURL,
			string runID)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("userId", _user.userId));
			apiClient.Queries.Add(new KeyValue("projectId", projectID));
			apiClient.Queries.Add(new KeyValue("slackURL", slackURL));
			apiClient.Queries.Add(new KeyValue("ProjectName", projectName));
			apiClient.Queries.Add(new KeyValue("type", projectType));
			apiClient.Queries.Add(new KeyValue("QIFURL", qvizURL));
			apiClient.Queries.Add(new KeyValue("runId", runID));
			apiClient.Body = testResult;
			apiClient.Post<object>("/api/integration/Slack");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		/// <summary>
		/// Send the provided GUI/API Testing Results to the JIRA Instance configured with QViz
		/// </summary>
		/// <param name="testResult">GUI or API Test Results as JSON Object</param>
		/// <param name="jiraURL">URL of the JIRA REST v2 API Instance</param>
		/// <param name="userName">Unique Name of the JIRA User</param>
		/// <param name="apiToken">JIRA REST v2 API Access Token of the User</param>
		public void SendTestResultsToJIRA(
			object testResult,
			string jiraURL,
			string userName,
			string apiToken)
		{
			var apiClient = new APIClient(_baseURL);
			apiClient.Headers.Add(new KeyValue("Content-Type", "application/json"));
			apiClient.Headers.Add(new KeyValue("Authorization", "Bearer " + _user.access_token));
			apiClient.Queries.Add(new KeyValue("jiraAPIURL", jiraURL));
			apiClient.Queries.Add(new KeyValue("username", userName));
			apiClient.Queries.Add(new KeyValue("passwordKey", apiToken));
			apiClient.Queries.Add(new KeyValue("postComment", "true"));
			apiClient.Queries.Add(new KeyValue("logDefect", "true"));
			apiClient.Body = testResult;
			apiClient.Post<object>("/api/integration/Jira");
			_lastResponse = apiClient.Response.Content;
			if (!apiClient.Response.IsSuccessful) throw apiClient.GetError();
		}

		#endregion
	}
}

File: QViz.Java/src/main/java/com/periscope/qviz/client/QVizClient.java
package com.periscope.qviz.client;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.azure.storage.CloudStorageAccount;
import com.microsoft.azure.storage.blob.CloudBlobClient;
import com.microsoft.azure.storage.blob.CloudBlobContainer;
import com.microsoft.azure.storage.blob.CloudBlockBlob;
import com.periscope.qviz.json.*;
import org.apache.http.util.EntityUtils;

import java.io.File;
import java.io.FileInputStream;
import java.net.URI;
import java.text.SimpleDateFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Client Library for interfacing with QViz APIs
 *
 * <script src="https://cdn.jsdelivr.net/gh/google/code-prettify/loader/run_prettify.js"></script>
 */
public class QVizClient {

	private URI baseURL;
	private AccessUser user;
	private ObjectMapper jsonMapper;
	private String lastResponse;

	/**
	 * Initialize QViz Client with the default Base URL for the API Instance of QViz
	 *
	 * @throws Exception In case of any failures
	 */
	public QVizClient() throws Exception {
		this.baseURL = new URI("https://api.qviz.io");
		initialize();
	}

	/**
	 * Initialize QViz Client with the provided Base URL for the API Instance of QViz
	 *
	 * @param apiURL Base URL to the API Instance of QViz
	 * @throws Exception In case of any failures
	 */
	public QVizClient(String apiURL) throws Exception {
		this.baseURL = new URI(apiURL);
		initialize();
	}

	/**
	 * Initialize QViz Client with the provided Base URL
	 */
	private void initialize() {
		SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS");
		this.jsonMapper = new ObjectMapper();
		this.jsonMapper.setDateFormat(dateFormat);
	}

	/**
	 * Get the Response from the QViz API that was executed last
	 *
	 * @return Text content of the Response from the QViz API that was executed last
	 */
	public String getLastResponse() {
		return lastResponse;
	}

	/**
	 * Authenticate with QViz using the provided User credentials
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     // Successful authentication
	 * } catch(Exception error) {
	 *     // Failed authentication
	 *     throw error;
	 * }</pre>
	 *
	 * @param userName Name of the User
	 * @param password Password of the User
	 * @throws Exception In case of failure
	 */
	public void authenticate(String userName, String password) throws Exception {
		Authentication authentication = new Authentication(userName, password);
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.body = jsonMapper.writeValueAsString(authentication);
		if (apiClient.post("/api/authentication") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
			this.user = jsonMapper.readValue(lastResponse, AccessUser.class);
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Get the list of Projects associated with the User
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     List{@code <Project>} projects = client.getProjects();
	 *     // Successfully retrieved the list of Projects
	 * } catch(Exception error) {
	 *     // Failed to get the list of Projects
	 *     throw error;
	 * }</pre>
	 *
	 * @return List of Projects as Array of Objects
	 * @throws Exception In case of failure
	 */
	public List<Project> getProjects() throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("userId", this.user.userId));
		if (apiClient.get("/api/projects") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
			return jsonMapper.readValue(lastResponse, new TypeReference<List<Project>>() {
			});
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Get a specific Project associated with the User
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     Project testProject = client.getProject("Test Project");
	 *     // Successfully retrieved the Project
	 * } catch(Exception error) {
	 *     // Failed to get the Project
	 *     throw error;
	 * }</pre>
	 *
	 * @param projectName Name of the Project to be fetched
	 * @return Project Object
	 * @throws Exception In case of failure
	 */
	public Project getProject(String projectName) throws Exception {
		List<Project> projects = this.getProjects();
		for (Project project : projects) {
			if (project.projectName.equals(projectName)) {
				return project;
			}
		}
		// Throw an Exception if the given Project Name was not found.
		throw new Exception(
			"Project (" + projectName + ") was not found in the User (" + this.user.username + ") account."
		);
	}

	/**
	 * Get the list of Configurations associated with the give Project
	 *
	 * @param projectId Identifier of the Project to get Configurations from
	 * @return List of Configurations for the given Project
	 * @throws Exception In case of failure
	 */
	public List<ProjectConfiguration> getProjectConfigurations(String projectId) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.headers.add(new KeyValue("userId", this.user.userId));
		apiClient.queries.add(new KeyValue("projectId", projectId));
		if (apiClient.get("/api/projectConfigurations") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
			return jsonMapper.readValue(lastResponse, new TypeReference<List<ProjectConfiguration>>() {
			});
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Get a specific Tool Configuration from the Project Configurations
	 *
	 * @param projectId Identifier of the Project
	 * @param toolName  Name of the Tool (SLACK, JIRA, TESTRAIL, etc.)
	 * @return Tool Configuration Data as Linked Hash Map of JSON Key and Value pairs
	 * @throws Exception In case of failure
	 */
	public LinkedHashMap<String, String> getToolConfiguration(String projectId, String toolName) throws Exception {
		List<ProjectConfiguration> projectConfigurations = this.getProjectConfigurations(projectId);
		for (ProjectConfiguration configuration : projectConfigurations) {
			if (configuration.value.indexOf(toolName) > 0) {
				return jsonMapper.readValue(configuration.value,
					new TypeReference<LinkedHashMap<String, String>>() {
					});
			}
		}
		// Throw an Exception if the given Tool Name was not found
		throw new Exception(
			"Project (" + projectId + ") does not contain a Tool Configuration for (" + toolName + ") tool."
		);
	}

	/**
	 * Get the list of GUI Test Cases associated with the specified Project
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     Project guiProject = client.getProject("GUI Test Project");
	 *     List{@code <TestCaseGUI>} guiTestCases = client.getGUITestCases(guiProject.projectId, false, "");
	 *     // Successfully retrieved the list of GUI Test Cases
	 * } catch(Exception error) {
	 *     // Failed to get the list of GUI Test Cases
	 *     throw error;
	 * }</pre>
	 *
	 * @param projectId       Identifier for the Project to fetch GUI Test Cases from
	 * @param onlyAutomated   Get only the GUI Test Cases that are Automated?
	 * @param onlyDescription Get only the GUI Test Case(s) that matches the Description?
	 * @return List of GUI Test Cases as Array of Objects
	 * @throws Exception In case of failure
	 */
	public List<TestCaseGUI> getGUITestCases(
		String projectId,
		boolean onlyAutomated,
		String onlyDescription
	) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("projectId", projectId));
		apiClient.headers.add(new KeyValue("userId",this.user.userId));
		if (apiClient.get("/api/uiTestcases") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
			List<TestCaseGUI> guiCases = jsonMapper.readValue(lastResponse, new TypeReference<List<TestCaseGUI>>() {
			});
			// Filter only Automated Test Cases (if requested)
			if (onlyAutomated) {
				guiCases = guiCases.stream().filter(
					tc -> tc.isAutomated.equals(true)).collect(Collectors.toList()
				);
			}
			// Filter only Test Cases that matches the Description (if requested)
			if (onlyDescription != null && !onlyDescription.isEmpty()) {
				guiCases = guiCases.stream().filter(
					tc -> tc.description.equals(onlyDescription)).collect(Collectors.toList()
				);
			}
			return guiCases;
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Get the list of API Test Cases associated with the specified Project
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     Project apiProject = client.getProject("API Test Project");
	 *     List{@code <TestCaseAPI>} apiTestCases = client.getAPITestCases(apiProject.projectId, false, "");
	 *     // Successfully retrieved the list of API Test Cases
	 * } catch(Exception error) {
	 *     // Failed to get the list of API Test Cases
	 *     throw error;
	 * }</pre>
	 *
	 * @param projectId       Identifier for the Project to fetch API Test Cases from
	 * @param onlyAutomated   Get only the API Test Cases that are Automated?
	 * @param onlyDescription Get only the API Test Case(s) that matches the Description?
	 * @return List of API Test Cases as Array of Objects
	 * @throws Exception In case of failure
	 */
	public List<TestCaseAPI> getAPITestCases(
		String projectId,
		boolean onlyAutomated,
		String onlyDescription
	) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("projectId", projectId));
		apiClient.headers.add(new KeyValue("userId", user.userId));
		if (apiClient.get("/api/apiTestcases") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
			List<TestCaseAPI> apiCases = jsonMapper.readValue(lastResponse, new TypeReference<List<TestCaseAPI>>() {
			});
			// Filter only Automated Test Cases (if requested)
			if (onlyAutomated) {
				apiCases = apiCases.stream().filter(
					tc -> tc.isAutomated.equals(true)).collect(Collectors.toList()
				);
			}
			// Filter only Test Cases that matches the Description (if requested)
			if (onlyDescription != null && !onlyDescription.isEmpty()) {
				apiCases = apiCases.stream().filter(
					tc -> tc.description.equals(onlyDescription)).collect(Collectors.toList()
				);
			}
			return apiCases;
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Upload the given Screen Shot File into Azure BLOB Storage and return the URL
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     File file = new File("files/screenshot_date_time.png");
	 *     String screenShotURL = client.uploadScreenShot("blob-connection-string", "blob-container-name", file);
	 *     // Successfully uploaded the Screen Shot and retrieved the BLOB URL
	 * } catch(Exception error) {
	 *     // Failed to upload the Screen Shot
	 *     throw error;
	 * }</pre>
	 *
	 * @param connection Azure Storage Account Connection String
	 * @param container  Azure Storage Container Name
	 * @param file       Screen Shot File
	 * @return Full URL of the Screen Shot File from Azure BLOB Storage
	 */
	public String uploadScreenShot(String connection, String container, File file) {
		try {
			if (!file.exists()) {
				throw new Exception("Provided Screen Shot File does not exist.");
			}
			CloudStorageAccount account = CloudStorageAccount.parse(connection);
			CloudBlobClient blobClient = account.createCloudBlobClient();
			CloudBlobContainer blobContainer = blobClient.getContainerReference(container);
			CloudBlockBlob blob = blobContainer.getBlockBlobReference(file.getName());
			blob.upload(new FileInputStream(file), file.length());
			return blob.getUri().toString();
		} catch (Exception error) {
			return "";
		}
	}

	/**
	 * Post the provided GUI Testing Results to the QViz Instance
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     GUITestResult guiTestResult = new GUITestResult();
	 *     // Update the Properties of guiTestResult object before posting
	 *     client.postGUITestResults(guiTestResult);
	 *     // Successfully posted the GUI Test Results
	 * } catch(Exception error) {
	 *     // Failed to post the GUI Test Results
	 *     throw error;
	 * }</pre>
	 *
	 * @param guiTestResult GUI Test Results as JSON Object
	 * @throws Exception In case of failure
	 */
	public void postGUITestResults(GUITestResult guiTestResult) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.headers.add(new KeyValue("userId", this.user.userId));
		apiClient.body = jsonMapper.writeValueAsString(guiTestResult);
		if (apiClient.post("/api/ui-results") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Post the provided API Testing Results to the QViz Instance
	 *
	 * <pre class="prettyprint">
	 * try {
	 *     QVizClient client = new QVizClient();
	 *     client.authenticate("user", "password");
	 *     APITestResult apiTestResult = new APITestResult();
	 *     // Update the Properties of apiTestResult object before posting
	 *     client.postAPITestResults(apiTestResult);
	 *     // Successfully posted the API Test Results
	 * } catch(Exception error) {
	 *     // Failed to post the API Test Results
	 *     throw error;
	 * }</pre>
	 *
	 * @param apiTestResult API Test Results as JSON Object
	 * @throws Exception In case of failure
	 */
	public void postAPITestResults(APITestResult apiTestResult) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.headers.add(new KeyValue("userId", this.user.userId));
		apiClient.body = jsonMapper.writeValueAsString(apiTestResult);
		if (apiClient.post("/api/api-results") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Send the provided GUI/API Testing Results to the TestRail Instance configured with QViz
	 *
	 * @param testResult    GUI or API Test Results as JSON Object
	 * @param testRailURL   URL of the TestRail Instance
	 * @param userName      Unique Name of the TestRail User
	 * @param apiToken      API Access Key of the TestRail User
	 * @param testRailRunID Identifier for the Test Run from TestRail Instance
	 * @param lastError     Last Error Message (if any)
	 * @throws Exception In case of failure
	 */
	public void sendTestResultsToTestRail(
		Object testResult,
		String testRailURL,
		String userName,
		String apiToken,
		String testRailRunID,
		String lastError
	) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("TestRailURL", testRailURL));
		apiClient.queries.add(new KeyValue("username", userName));
		apiClient.queries.add(new KeyValue("passwordKey", apiToken));
		apiClient.queries.add(new KeyValue("postResult", "true"));
		apiClient.queries.add(new KeyValue("TestRailRunID", testRailRunID));
		if (lastError.startsWith("Defect")) {
			apiClient.queries.add(
				new KeyValue(
					"Defect",
					lastError.split(": ")[1].split("::")[0].replace(" :", "")
				)
			);
		}
		apiClient.body = jsonMapper.writeValueAsString(testResult);
		if (apiClient.post("/api/integration/TestRail") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Send the provided GUI/API Testing Results to the Slack Instance configured with QViz
	 *
	 * @param testResult  GUI or API Test Results as JSON Object
	 * @param slackURL    URL of the Slack Hooks for sending Messages
	 * @param projectID   Identifier for the Project
	 * @param projectName Name of the Project
	 * @param projectType Type of the Project
	 * @param qvizURL     URL to the QViz App Instance
	 * @param runID       Identifier for the Run
	 * @throws Exception In case of failure
	 */
	public void sendTestResultsToSlack(
		Object testResult,
		String slackURL,
		String projectID,
		String projectName,
		String projectType,
		String qvizURL,
		String runID
	) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("userId", this.user.userId));
		apiClient.queries.add(new KeyValue("projectId", projectID));
		apiClient.queries.add(new KeyValue("slackURL", slackURL));
		apiClient.queries.add(new KeyValue("ProjectName", projectName));
		apiClient.queries.add(new KeyValue("type", projectType));
		apiClient.queries.add(new KeyValue("QIFURL", qvizURL));
		apiClient.queries.add(new KeyValue("runId", runID));
		apiClient.body = jsonMapper.writeValueAsString(testResult);
		if (apiClient.post("/api/integration/Slack") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
		} else {
			throw apiClient.getError();
		}
	}

	/**
	 * Send the provided GUI/API Testing Results to the JIRA Instance configured with QViz
	 *
	 * @param testResult GUI or API Test Results as JSON Object
	 * @param jiraURL    URL of the JIRA REST v2 API Instance
	 * @param userName   Unique Name of the JIRA User
	 * @param apiToken   JIRA REST v2 API Access Token of the User
	 * @throws Exception In case of failures
	 */
	public void sendTestResultsToJIRA(
		Object testResult,
		String jiraURL,
		String userName,
		String apiToken
	) throws Exception {
		APIClient apiClient = new APIClient(this.baseURL);
		apiClient.headers.add(new KeyValue("Content-Type", "application/json"));
		apiClient.headers.add(new KeyValue("Authorization", "Bearer " + this.user.access_token));
		apiClient.queries.add(new KeyValue("jiraAPIURL", jiraURL));
		apiClient.queries.add(new KeyValue("username", userName));
		apiClient.queries.add(new KeyValue("passwordKey", apiToken));
		apiClient.queries.add(new KeyValue("postComment", "true"));
		apiClient.queries.add(new KeyValue("logDefect", "true"));
		apiClient.body = jsonMapper.writeValueAsString(testResult);
		if (apiClient.post("/api/integration/Jira") == 200) {
			// DEBUG: Store the API Response
			lastResponse = EntityUtils.toString(apiClient.response.getEntity());
		} else {
			throw apiClient.getError();
		}
	}

}
